**Цель работы:** <br>
Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

## Загрузка набора данных

### Описание предметной области
Вариант № 23 (8) <br>
Набор данных: credit_risk.csv <br>
Атрибуты:
1. возраст
2. доход
3. статус домовладения (Собственный, Ипотека, Аренда)
4. стаж работы в годах
5. цель кредита
6. сумма кредита по заявке
7. процентная ставка
8. статус одобрения кредита
9. сумма кредита в процентах от дохода
10. не допустил ли заявитель ранее дефолта по кредиту (да, нет)
11. длина кредитной истории заявителя

#### 1. Чтение файла (набора данных)

In [5]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

#### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода `head`.

Vетод `head()` – по умолчанию выводит первые 5 строк набора данных, если в скобках
указать параметр, то возвращает количество строк, указанное в параметре.

In [6]:
import pandas as pd # импорт библиотеки

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8") # чтение таблицы, где url – путь к файлу
df.head(20)

Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100     RENET         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

Анализируя таблицу видно, что проблем с названиями столбцов нет: 
- в названиях нет пробелов, они разделены нижним подчёркиванием;
- в названиях используется один язык;
- каждое название отражает суть своего столбца.

2.2 Оценка данных с помощью метода info.

Метод `info()` , предоставляет общую
информацию о таблице, включая названия столбцов, типы данных, хранящиеся в них, и
количество ненулевых значений в каждом столбце.

In [11]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.3+ KB


2.3 Оценка данных с помощью метода describe.

In [12]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")
df.describe()

Id         Age         Income  Emp_length        Amount  \
count  652.000000  652.000000     652.000000  643.000000    652.000000   
mean   325.495399   24.285276   90008.855828    4.601866  18800.996933   
std    188.352535    7.841061   69530.312525    7.281616   9152.896250   
min      0.000000   21.000000    9600.000000    0.000000   1000.000000   
25%    162.750000   23.000000   44000.000000    2.000000  10000.000000   
50%    325.500000   24.000000   69998.000000    4.000000  21850.000000   
75%    488.250000   25.000000  128499.000000    7.000000  25000.000000   
max    649.000000  144.000000  500000.000000  123.000000  35000.000000   

             Rate      Status  Percent_income  Cred_length  
count  586.000000  652.000000      652.000000   652.000000  
mean    12.289471    0.601227        0.279831     3.012270  
std      3.272772    0.490022        0.150231     0.808843  
min      5.420000    0.000000        0.010000     2.000000  
25%     10.250000    0.000000        0.160000     2.000000  
50%     12.180000    1.000000        0.280000     3.000000  
75%     14.720000    1.000000        0.380000     4.000000  
max     21.210000    1.000000        0.830000     4.000000

2.4 Оценка названий столбцов

In [9]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")
df.columns

Index(['Id', 'Age', 'Income', 'Home', 'Emp_length', 'Intent', 'Amount', 'Rate',
       'Status', 'Percent_income', 'Default', 'Cred_length'],
      dtype='object')

#### 3. Проверка пропусков

Проверить данные на наличие пропусков и устранить их, если они есть

In [7]:
print(df.isna().sum())

Id                 0
Age                0
Income             0
Home               0
Emp_length         9
Intent             0
Amount             0
Rate              66
Status             0
Percent_income     0
Default            0
Cred_length        0
dtype: int64


Удаляем данные с пропуском

In [7]:
df = df.dropna(subset=['Emp_length', 'Rate'])

Проверяем, что данных с пропуском больше нет

In [8]:
print(df.isna().sum())

Id                 0
Age                0
Income             0
Home               0
Emp_length         9
Intent             0
Amount             0
Rate              66
Status             0
Percent_income     0
Default            0
Cred_length        0
dtype: int64


#### 4. Проверка дубликатов

4.1 Проверка явных дубликатов

In [23]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

print("\nДо удаления дубликатов")
print("Количество дубликатов:", df.duplicated().sum())

print("\nДубликаты строк:")
print(df[df.duplicated()])

df = df.drop_duplicates().reset_index(drop=True)

print("\nПосле удаления дубликатов")
print("Количество дубликатов:", df.duplicated().sum())


До удаления дубликатов
Количество дубликатов: 2

Дубликаты строк:
      Id   Age  Income  Home  Emp_length             Intent  Amount  Rate  \
650  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   
651  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   

     Status  Percent_income Default  Cred_length  
650       1            0.34       N            3  
651       1            0.34       N            3  

После удаления дубликатов
Количество дубликатов: 0


4.2 Проверка неявных дубликатов

In [6]:
import pandas as pd

# Загрузка данных
# df = pd.read_csv("C:/ГУАП/5 семестр/ВвАД/LR1/credit_risk.csv", encoding="utf-8")

print(df['Home'].value_counts())

df['Home'] = df['Home'].replace('OWN', 'Own').replace('MORTGAGE', 'Mortgage').replace('RENT', 'Rent').replace('RENET', 'Rent')

print(df['Home'].value_counts())

Home
Rent        415
Mortgage    136
Own          98
OTHER         2
RENET         1
Name: count, dtype: int64
Home
Rent        416
Mortgage    136
Own          98
OTHER         2
Name: count, dtype: int64


#### 5. Провека типов данных

In [28]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

print("До преобразования:\n", df.dtypes)

df["Id"] = df["Id"].astype("int64")
df["Age"] = df["Age"].astype("int64")          
df["Income"] = df["Income"].astype("int64")    
df["Home"] = df["Home"].astype("object")       
df["Emp_length"] = df["Emp_length"].astype("float64")  
df["Intent"] = df["Intent"].astype("object")   
df["Amount"] = df["Amount"].astype("int64")    
df["Rate"] = df["Rate"].astype("float64")     
df["Status"] = df["Status"].astype("int64")    
df["Percent_income"] = df["Percent_income"].astype("float64")
df["Cred_length"] = df["Cred_length"].astype("int64")

print("\nПосле преобразования:\n", df.dtypes)

До преобразования:
 Id                  int64
Age               float64
Income              int64
Home               object
Emp_length        float64
Intent             object
Amount              int64
Rate              float64
Status              int64
Percent_income    float64
Default            object
Cred_length         int64
dtype: object

После преобразования:
 Id                  int64
Age                 int64
Income              int64
Home               object
Emp_length        float64
Intent             object
Amount              int64
Rate              float64
Status              int64
Percent_income    float64
Default            object
Cred_length         int64
dtype: object


#### 6. Группировка данных

Задание 1 <br>
Группировка - статус домовладения (home) по количеству дефолтов
(default) . Результат должен иметь примерно следующий вид.

In [9]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

result = df.groupby("Home")["Default"].value_counts()

print(result)

Home      Default
MORTGAGE  N          107
          Y           29
OTHER     Y            2
OWN       N           85
          Y           13
RENET     N            1
RENT      N          306
          Y          106
          No           3
Name: count, dtype: int64


**Вывод:** Из результата группировки видно распределение дефолтов (Default) по статусу домовладения (Home). Наибольшее количество заявителей с кредитами — у арендаторов (RENT), среди них встречается как наибольшее число без дефолта, так и значительная доля с дефолтом.

Задание 2 <br>
Группировка - цель кредита (intent) по количеству статуса
домовладения (home). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”. 

In [11]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

grouped = df.groupby('Home')['Intent'].value_counts().reset_index(name="count")

grouped = grouped.sort_values(by='count', ascending=False)

print(grouped)

        Home             Intent  count
15      RENT          EDUCATION     93
16      RENT  DEBTCONSOLIDATION     74
17      RENT            MEDICAL     70
18      RENT            VENTURE     70
19      RENT           PERSONAL     69
20      RENT    HOMEIMPROVEMENT     39
0   MORTGAGE          EDUCATION     34
1   MORTGAGE            VENTURE     27
2   MORTGAGE            MEDICAL     22
8        OWN           PERSONAL     22
3   MORTGAGE  DEBTCONSOLIDATION     21
9        OWN            VENTURE     20
4   MORTGAGE           PERSONAL     19
11       OWN    HOMEIMPROVEMENT     16
10       OWN          EDUCATION     16
12       OWN            MEDICAL     15
5   MORTGAGE    HOMEIMPROVEMENT     13
13       OWN  DEBTCONSOLIDATION      9
7      OTHER           PERSONAL      1
6      OTHER          EDUCATION      1
14     RENET          EDUCATION      1


**Вывод:** Группировка показала, что основная часть заявок приходится на RENT, где лидируют цели EDUCATION, DEBTCONSOLIDATION, MEDICAL, VENTURE и PERSONAL. Для MORTGAGE преобладают EDUCATION и VENTURE, а для OWN — PERSONAL и HOMEIMPROVEMENT. Встречаются единичные значения для категорий OTHER и RENET.

Задание 3 <br>
Сводная таблица (pivot_table) - средний доход income для каждого
возраста. Отсортировать по убыванию. Округлить до двух знаков.

In [46]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")

pivot = df.pivot_table(values='Income', index='Age', aggfunc='mean')

pivot = pivot.sort_values(by='Income', ascending=False)

pivot = pivot.round(2)

print(pivot)


          Income
Age             
144.0  225000.00
26.0   122793.74
25.0   106750.02
24.0    88684.03
23.0    85275.05
123.0   80004.00
22.0    64423.88
21.0    42238.58


**Вывод:** Сводная таблица показывает средний доход по возрастам. Наибольшие значения приходятся на возрасты 144 и 123 лет, что указывает на выбросы в данных. Среди реальных возрастов самый высокий средний доход наблюдается у группы 26 лет, далее идут 25, 24 и 23 года.

Задание 4 <br>
Сводная таблица (pivot_table) - средняя сумма кредита amount для
каждой цели (intent) - строки и возраста - столбцы . Отсортировать по
возрастанию intent. Округлить до двух знаков.

In [4]:
import pandas as pd

url = r"C:\ГУАП\5 семестр\ВвАД\LR1\credit_risk.csv"
df = pd.read_csv(url, encoding="utf-8")


pivot = df.pivot_table(
    values='Amount',      
    index='Intent',       
    columns='Age',        
    aggfunc='mean'        
)

pivot = pivot.sort_index(ascending=True)

pivot = pivot.round(2)

print(pivot)


Age                   21.0      22.0      23.0      24.0      25.0      26.0   \
Intent                                                                          
DEBTCONSOLIDATION   9500.00  18602.78  20275.00  20028.57  18603.33  21947.22   
EDUCATION          12846.15  18677.34  20288.28  21052.38  19326.04  20566.67   
HOMEIMPROVEMENT     4355.00  10216.67   4000.00  19627.68  21266.67  20038.46   
MEDICAL            10700.00  18073.21  17968.48  21119.64  18114.81  20410.00   
PERSONAL            3633.33  17735.71  19241.00  21205.00  21738.00  18588.24   
VENTURE             9407.69  15515.28  22810.94  19406.73  18360.42  18850.00   

Age                  123.0   144.0  
Intent                              
DEBTCONSOLIDATION      NaN     NaN  
EDUCATION          20400.0  6000.0  
HOMEIMPROVEMENT        NaN     NaN  
MEDICAL                NaN     NaN  
PERSONAL               NaN     NaN  
VENTURE                NaN  4800.0  


**Вывод:** Сводная таблица показывает средние суммы кредита по целям (Intent) и возрастам (Age). Наибольшие значения встречаются у категорий EDUCATION, MEDICAL и VENTURE в возрастах 23–25 лет, что говорит о повышенной сумме заявок в этих группах.

#### Вывод

В ходе работы была проведена предварительная обработка данных CSV-файла, в процессе которой изучались основные атрибуты датасета, проверялись и устранялись возможные проблемы, такие как пропущенные либо некорректные значения и выбросы. Далее был выполнен анализ категориальных признаков и проведена нормализация отдельных полей для обеспечения корректности группировок. Для более глубокого изучения информации была построена сводная таблица, отразившая средние суммы кредитов в зависимости от целей кредита (Intent) и возрастных групп клиентов (Age).